### DISCUSSION TITLE
[Next 13] Why do I have to mark this page as client component?

### DISCUSSION THREAD

**Discussion by nareshbhatia** [vercel/next.js#46227] [Help] [answered]
I am following the guidelines listed under [Server and Client Components](https://beta.nextjs.org/docs/rendering/server-and-client-components), but still getting this unhandled runtime error on a page: "Unsupported Server Component type: undefined". If I make this page a client component, the error goes away. Here's my component hierarchy:

<img width="1299" alt="Movie Magic Component Hierarchy" src="https://user-images.githubusercontent.com/1220198/220537633-f7237500-b62b-45db-8012-c2803cfe10cb.png">

The page in question is `Top10MoviesPage`. It doesn't do anything special that should force me to mark it as a client component. The child component `Header` is marked as client component because it is interactive. The child component `Top10MovieList` is a server component that fetches a list of movies and renders it. Here's the code for `Top10MoviesPage`:

```tsx
'use client';

import * as React from 'react';
import { Header } from '~/components';
import { Top10MoviesList } from './Top10MoviesList';

export default function Top10MoviesPage() {
  return (
    <div className="mx-auto max-w-7xl">
      <Header />
      <div className="mx-auto max-w-3xl px-4 py-8">
        <h1 className="text-center text-3xl font-bold leading-tight tracking-tight">
          Top 10 Movies Of All Time
        </h1>

        <div className="mt-6">
          <Top10MoviesList />
        </div>
      </div>
    </div>
  );
}
```

### Questions:
1. How can I remove the `use client` from `Top10MoviesPage`?
2. I am seeing that the Top10MovieList component is fetching data on the server side as well as the client side. Why? I was expecting it to only fetch on the server side and render the list.

Any insights would be very helpful.

My full source code is available here: https://github.com/nareshbhatia/movie-magic-2

**Comment by icyJoseph** [comment, marked as answer]
> // <Top10MoviesList> should fetch data only on the
> //  server side, but it is also fetching data on the

What makes you think that? `React.use` is a client side hook.

From the beta docs:

> This new data fetching model is currently being developed by the React team. We recommend reading the [support for promises React RFC](https://github.com/acdlite/rfcs/blob/first-class-promises/text/0000-first-class-support-for-promises.md) which introduces async/await in Server Components and a new use() hook for client components.

To fetch server side only, you gotta turn `Top10MoviesPage` into a Server component by removing `React.use` and simply do async/await:

```js
export async function Top10MoviesList() {
  const movies = await getMovies();
```

Making a stretch here but the `Suspense Boundary` + `use` is a client side replacement for `useState + useEffect` handling of promises, while keeping Client Components as sync functions. 

Btw, also from the docs (a reminder that this is beta technology to play, test, and understand the intention, but not to master just yet):

> Wrapping fetch in use is currently not recommended in Client Components and may trigger multiple re-renders. For now, if you need to fetch data in a Client Component, we recommend using a third-party library such as [SWR](https://swr.vercel.app/) or [React Query](https://tanstack.com/query/v4).
>
> Note: We'll be adding more examples once fetch and use work in Client Components.

**Reply by nareshbhatia** [reply]
@icyJoseph, thank you for clarifying a bunch of concepts and some great references!

> `should fetch data only on the server side, but it is also fetching data on the client side`
>
> What makes you think that?

Because the `console.log` inside `getMovies()` is logging on both server and client.

> To fetch server side only, you gotta turn Top10MoviesPage into a Server component by removing React.use and simply do async/await

Tried this, but it does not work. The `console.log` is not logging on the server side at all. On client side I am getting this error:

> Unhandled Runtime Error
> Error: Objects are not valid as a React child (found: [object Promise]). If you meant to render a collection of children, use an array instead.

Here's the modified code:

```tsx
export async function Top10MoviesList() {
  const movies = await getMovies();
  return (
    <ul className="flex-none space-y-4">
      ...
    </ul>
  );
}
```

So the basic issue is that React thinks this component is a client component. Need to figure out how to make it a true server side component.

Looking at the component hierarchy above, it seems that React is treating a component as a client component if its parent is a client component.

**Reply by icyJoseph** [reply]
Right, according to https://beta.nextjs.org/docs/rendering/server-and-client-components#importing-server-components-into-client-components:

```js
// ❌ This pattern will not work. You cannot import a Server
// Component into a Client Component
```

I was expecting you'd remove the "use client" from your page level component, `src/app/top-10-mo1vies/page.tsx`, because only Header needs client side interactivity, and that one already has "use client" directive applied.

You could also just use composition, https://reactjs.org/docs/composition-vs-inheritance.html, using the children prop, as the link above shows, and compose the Client and Server Components through a 3rd Server Component. I know this is a bit convoluted, but it is what it is.

**Reply by nareshbhatia** [reply]
Hi @icyJoseph, simply removing "use client" from `Top10MoviesPage` does not work. React still treats it as a client component because it is a child of the `AuthStateContextProvider` which is a client component. So I left the "use client" in `Top10MoviesPage`. That way I could try the other things below.

As you suggested, I have now wrapped `<Top10MoviesList>` with a client component and passing it in as a prop. However even this does not work, giving the same error: "Objects are not valid as a React child (found: [object Promise])". Here's the modified code:

```tsx
'use client';

import * as React from 'react';
import { Header } from '~/components';
import { Top10MoviesList } from './Top10MoviesList';

interface Top10MoviesListClientProps {
  children: React.ReactNode;
}

function Top10MoviesListClient({ children }: Top10MoviesListClientProps) {
  return <>{children}</>;
}

export default function Top10MoviesPage() {
  return (
    <div className="mx-auto max-w-7xl">
      <Header />
      <div className="mx-auto max-w-3xl px-4 py-8">
        <h1 className="text-center text-3xl font-bold leading-tight tracking-tight">
          Top 10 Movies Of All Time
        </h1>

        <div className="mt-6">
          <Top10MoviesListClient>
            <Top10MoviesList />
          </Top10MoviesListClient>
        </div>
      </div>
    </div>
  );
}
```

This is turning out to be more complicated than I thought. Any other ideas? I have checked in the currently broken code in my repo (https://github.com/nareshbhatia/movie-magic-2). Please feel free to give it a try. (Also tagging @leerob.)

As an aside, the `<Top10MoviesList />` line above is giving a TypeScript error. It is upset about this component being async and returning `Promise<Element>` instead of `Element`. Wonder if anyone has hit this one:

> TS2786: 'Top10MoviesList' cannot be used as a JSX component.
> Its return type 'Promise<Element>' is not a valid JSX element.
> Type 'Promise<Element>' is missing the following properties from type 'ReactElement<any, any>': type, props, key

**Reply by icyJoseph** [reply]
Regarding the TypeScript error that's something that we fix like this:

```tsx
<SomeJSX>
  {/* @ts-expect-error Server Component */}
   <SomeAsyncAwaitComponent />
</SomeJSX>
```

From the docs, https://beta.nextjs.org/docs/configuring/typescript:

> Warning: You can use async/await in layouts and pages, which are Server Components. Using async/await inside other components, with TypeScript, can cause errors from the response type from JSX. You might see the error 'Promise<Element>' is not a valid JSX element. We are tracking this [issue here](https://github.com/vercel/next.js/issues/42292).
> 
> We are working with the TypeScript team to resolve this upstream. As a temporary workaround, you can use {/* @ts-expect-error Server Component */} to disable type checking for the component.

Once the issue is fixed upstream, TS will complain about a useless `ts-expect-error` directive, and you can just remove it.

**Reply by icyJoseph** [reply]
I am almost tempted to explore your repository, but as said earlier, these are things to try and play around with, but striving to master them as they are right now is perhaps not the best time investment. 

> This is turning out to be more complicated than I thought.

Maybe this project is too large to try and move everything in one go. As said above, I am tempted to clone it but, maybe not...

I follow closely and read as much as I can about patterns and intentions, and every now and then I try to migrate some random app to it, but as soon as I hit a road block, I search to see if others have, if not I make an issue and move on.

**Reply by icyJoseph** [reply]
Alright, you [nerd-sniped me](https://xkcd.com/356/)... 

![Screenshot 2023-02-22 at 19 36 36](https://user-images.githubusercontent.com/21013447/220725258-dedd7079-fb69-45bc-af28-ac348df5b36a.png)

```tsx
// import * as React from "react";
import { Header } from "~/components/Header";
import { Top10MoviesList } from "./Top10MoviesList";

export default function Top10MoviesPage() {
  return (
    <div className="mx-auto max-w-7xl">
      <Header />

      <div className="mx-auto max-w-3xl px-4 py-8">
        <h1 className="text-center text-3xl font-bold leading-tight tracking-tight">
          Top 10 Movies Of All Time
        </h1>

        <div className="mt-6">
          {/* @ts-expect-error Server Component */}
          <Top10MoviesList />
        </div>
      </div>
    </div>
  );
}
```

There's some issue with your Header import though, I had to change it to `import { Header } from "~/components/Header";`

I am also not a big fan of `import * as React from 'react';` but I guess that's bike shedding from my side.

**Reply by icyJoseph** [reply]
@nareshbhatia someone just had the same issue I experienced with your Header: https://github.com/vercel/next.js/discussions/46259#discussion-4887088 I guess after all there's something deeper going on here...

**Reply by nareshbhatia** [reply]
@icyJoseph, you are the best! Thank you so much for the help.

And in the process, I learned a new term: nerd-sniping :-)
